### ISSUE TITLE
Saltstack wrong init system while working with services

### ISSUE THREAD

**Issue by alexminder** [saltstack/salt#37433]
### Description of Issue/Question
```
----------
          ID: paxctld
    Function: service.running
      Result: False
     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/usr/lib64/python2.7/site-packages/salt/state.py", line 1733, in call                                                                                                                                                                                    
                  **cdata['kwargs'])                                                                                                                                                                                                                                           
                File "/usr/lib64/python2.7/site-packages/salt/loader.py", line 1652, in wrapper                                                                                                                                                                                
                  return f(*args, **kwargs)                                                                                                                                                                                                                                    
                File "/usr/lib64/python2.7/site-packages/salt/states/service.py", line 601, in mod_watch                                                                                                                                                                       
                  __salt__['service.stop'](name)                                                                                                                                                                                                                               
                File "/usr/lib64/python2.7/site-packages/salt/modules/gentoo_service.py", line 141, in stop                                                                                                                                                                    
                  return not __salt__['cmd.retcode'](cmd, python_shell=False)                                                                                                                                                                                                  
                File "/usr/lib64/python2.7/site-packages/salt/modules/cmdmod.py", line 1773, in retcode                                                                                                                                                                        
                  password=kwargs.get('password', None))                                                                                                                                                                                                                       
                File "/usr/lib64/python2.7/site-packages/salt/modules/cmdmod.py", line 475, in _run                                                                                                                                                                            
                  'reason: {2}'.format(cmd, kwargs, exc)                                                                                                                                                                                                                       
              CommandExecutionError: Unable to run command '['/etc/init.d/paxctld', 'stop']' with the context '{'timeout': None, 'with_communicate': True, 'shell': False, 'bg': False, 'stderr': -2, 'env': {'LANG': 'POSIX', 'LC_NUMERIC': 'C', 'LC_MESSAGES': 'C', 'LC_IDENTIFICATION': 'C', 'LC_MONETARY': 'C', 'DEBIAN_FRONTEND': 'noninteractive', 'LC_COLLATE': 'C', 'LC_CTYPE': 'C', 'LC_ADDRESS': 'C', 'LC_MEASUREMENT': 'C', 'LC_TELEPHONE': 'C', 'LC_PAPER': 'C', 'LC_NAME': 'C', 'PATH': '/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin', 'LC_TIME': 'C', 'UCF_FORCE_CONFFOLD': '1', 'APT_LISTBUGS_FRONTEND': 'none', 'APT_LISTCHANGES_FRONTEND': 'none'}, 'stdout': -1, 'close_fds': True, 'stdin': None, 'cwd': '/root'}', reason: [Errno 2] No such file or directory                                          
     Started: 13:59:56.934705
    Duration: 77.733 ms
     Changes:   
```

Saltstack for Gentoo hosts with systemd try to use openrc/init-scripts.
Accordind to https://wiki.gentoo.org/wiki/Comparison_of_init_systems Systemd is second supported init system.

I think, Salt must use "init" grains to choose init system to use:

```
alex-desktop ~ # salt $(hostname -f) grains.get init
alex-desktop.home:
    systemd
```

### Setup
(Please provide relevant configs and/or SLS files (Be sure to remove sensitive info).)
```
paxctld:
   pkg:
      - installed
      - name: sys-apps/paxctld
   service:
      - running
      - enable: True
      - watch:
         - file: paxctld.conf

paxctld.conf:
  file.managed:
    - order: 1
    - name: /etc/paxctld.conf
    - user: root
    - group: root
    - mode: '0644'
    - source: salt://paxctld/files/paxctld.conf
    - require:
       - pkg: paxctld
```
### Steps to Reproduce Issue
(Include debug logs if possible and relevant.)
- Install Gentoo Gnu/Linux
- Select systemd profile and update world
- Install SaltStack
- Run state which use service control

### Versions Report
(Provided by running `salt --versions-report`. Please also mention any differences in master/minion versions.)
```
alex-desktop ~ # salt --versions-report
Salt Version:
           Salt: 2016.3.3
 
Dependency Versions:
           cffi: 1.5.2
       cherrypy: Not Installed
       dateutil: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: 0.22
           Mako: 1.0.0
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: 2.14
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.10 (default, May  6 2016, 14:15:19)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 14.4.1
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4
 
System Versions:
           dist: gentoo 2.2 
        machine: x86_64
        release: 4.7.10-hardened-r1
         system: Linux
        version: Gentoo Base System 2.2
```

**Comment by gtmanfred** [comment]
I am going to work on replicating this later today.

In the mean time, you should be able to set 

```
provider:
  service: systemd
```
in the minion config, maybe in the grains or pillars, and it will use the systemd module as the service provider for you.

Thanks,
Daniel

**Comment by gtmanfred** [comment]
Hrm, I am unable to replicate this.

I have openrc and systemd enabled, but systemd is the booted init system, and everything is working as expected.

I installed using the minimal iso, and configured it using the basic systemd profile.  Is there any other information you can give me about this?

**Comment by alexminder** [comment]
As I understand, check for systemd environment uses this code:

...salt/modules/systemd.py:
```
def __virtual__():
    '''
    Only work on systems that have been booted with systemd
    '''
    if __grains__['kernel'] == 'Linux' \
            and salt.utils.systemd.booted(__context__):
        return __virtualname__
    return (
        False,
        'The systemd execution module failed to load: only available on Linux '
        'systems which have been booted with systemd.'
    )
```
So, 'kernel' grains must be 'Linux' and "salt.utils.systemd.booted" must be 'True'.

How about "salt.utils.systemd.booted" ?.
...salt/utils/systemd.py:
```
def booted(context=None):
    '''
    Return True if the system was booted with systemd, False otherwise.
    Pass in the loader context "__context__", this function will set the
    systemd.sd_booted key to represent if systemd is running
    '''
    contextkey = 'salt.utils.systemd.booted'
    if isinstance(context, dict):
        # Can't put this if block on the same line as the above if block,
        # because it willl break the elif below.
        if contextkey in context:
            return context[contextkey]
    elif context is not None:
        raise SaltInvocationError('context must be a dictionary if passed')

    try:
        # This check does the same as sd_booted() from libsystemd-daemon:
        # http://www.freedesktop.org/software/systemd/man/sd_booted.html
        ret = bool(os.stat('/run/systemd/system'))
    except OSError:
        ret = False

    try:
        context[contextkey] = ret
    except TypeError:
        pass

    return ret
```
```
alex-desktop ~ # stat /run/systemd/system; echo $?
  File: '/run/systemd/system'
  Size: 80              Blocks: 0          IO Block: 4096   directory
Device: ch/12d  Inode: 9625        Links: 4
Access: (0755/drwxr-xr-x)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2016-11-05 23:24:29.685274056 +0300
Modify: 2016-11-05 23:20:38.209020664 +0300
Change: 2016-11-05 23:20:38.209020664 +0300
 Birth: -
0
```

I set provider option to the /etc/salt/minion and restartet salt-minion:
```
master: localhost
providers:
  service: systemd
```
With this option systemd module works fine:
```
alex-desktop ~ # salt-call -l debug service.restart paxctld
[DEBUG   ] Reading configuration from /etc/salt/minion
[DEBUG   ] Including configuration from '/etc/salt/minion.d/_schedule.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/_schedule.conf
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: alex-desktop.home
[DEBUG   ] Configuration file path: /etc/salt/minion
[WARNING ] Insecure logging configuration detected! Sensitive data may be logged.
[DEBUG   ] Reading configuration from /etc/salt/minion
[DEBUG   ] Including configuration from '/etc/salt/minion.d/_schedule.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/_schedule.conf
[DEBUG   ] Please install 'virt-what' to improve results of the 'virtual' grain.
[WARNING ] /usr/lib64/python2.7/site-packages/salt/grains/core.py:1493: DeprecationWarning: The "osmajorrelease" will be a type of an integer.

[DEBUG   ] Connecting to master. Attempt 1 of 1
[DEBUG   ] Initializing new AsyncAuth for ('/etc/salt/pki/minion', 'alex-desktop.home', 'tcp://127.0.0.1:4506')
[DEBUG   ] Generated random reconnect delay between '1000ms' and '11000ms' (4367)
[DEBUG   ] Setting zmq_reconnect_ivl to '4367ms'
[DEBUG   ] Setting zmq_reconnect_ivl_max to '11000ms'
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', 'alex-desktop.home', 'tcp://127.0.0.1:4506', 'clear')
[DEBUG   ] Decrypting the current master AES key
[DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
[DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
[DEBUG   ] Determining pillar cache
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', 'alex-desktop.home', 'tcp://127.0.0.1:4506', 'aes')
[DEBUG   ] Initializing new AsyncAuth for ('/etc/salt/pki/minion', 'alex-desktop.home', 'tcp://127.0.0.1:4506')
[DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
[DEBUG   ] LazyLoaded jinja.render
[DEBUG   ] LazyLoaded yaml.render
[DEBUG   ] LazyLoaded cmd.run
[INFO    ] Executing command ['systemctl', 'status', 'paxctld.service', '-n', '0'] in directory '/root'
[DEBUG   ] output: * paxctld.service - PaX flags maintenance daemon
   Loaded: loaded (/usr/lib64/systemd/system/paxctld.service; enabled; vendor preset: disabled)
   Active: active (running) since Sat 2016-11-05 23:17:22 MSK; 1h 5min ago
 Main PID: 1226 (paxctld)
   CGroup: /system.slice/paxctld.service
           `-1226 /sbin/paxctld
[INFO    ] Executing command ['systemctl', 'is-enabled', 'paxctld.service'] in directory '/root'
[DEBUG   ] output: enabled
[DEBUG   ] Service 'paxctld' is not masked
[DEBUG   ] LazyLoaded config.get
[INFO    ] Executing command ['systemd-run', '--scope', 'systemctl', 'restart', 'paxctld.service'] in directory '/root'

** (pkttyagent:7790): WARNING **: Unable to register authentication agent: GDBus.Error:org.freedesktop.PolicyKit1.Error.Failed: An authentication agent already exists for the given subject
Error registering authentication agent: GDBus.Error:org.freedesktop.PolicyKit1.Error.Failed: An authentication agent already exists for the given subject (polkit-error-quark, 0)
[DEBUG   ] output: Running scope as unit run-7784.scope.
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', 'alex-desktop.home', 'tcp://127.0.0.1:4506', 'aes')
[DEBUG   ] Initializing new AsyncAuth for ('/etc/salt/pki/minion', 'alex-desktop.home', 'tcp://127.0.0.1:4506')
[DEBUG   ] LazyLoaded nested.output
local:
    True
```

**Comment by alexminder** [comment]
If I broke the condition os=Gentoo, salt choose right init service:
salt/modules/gentoo_service.py:
```
def __virtual__():
    '''
    Only work on systems which default to systemd
    '''
    if __grains__['os'] == 'Gentoo':
        return __virtualname__
    return (False, 'The gentoo_service execution module cannot be loaded: '
            'only available on Gentoo systems.')
```
Very strange condition. It does not have check for systemd.

What we have for Debian systems:
salt/modules/debian_service.py:
```
def __virtual__():
    '''
    Only work on Debian and when systemd isn't running
    '''
    if __grains__['os'] in ('Debian', 'Raspbian', 'Devuan') and not salt.utils.systemd.booted(__context__):
        return __virtualname__
    else:
        return (False, 'The debian_service module could not be loaded: '
                'unsupported OS family and/or systemd running.')
```

So, I made the patch which fix the issue:
```
--- gentoo_service.py.orig      2016-11-06 00:52:20.090962306 +0300
+++ gentoo_service.py.new       2016-11-06 00:53:20.147461130 +0300
@@ -13,18 +13,21 @@
 # Import Python libs
 from __future__ import absolute_import
 
+# Import salt libs
+import salt.utils.systemd
+
 # Define the module's virtual name
 __virtualname__ = 'service'
 
 
 def __virtual__():
     '''
-    Only work on systems which default to systemd
+    Only work on Gentoo and when systemd isn't running
     '''
-    if __grains__['os'] == 'Gentoo':
+    if __grains__['os'] == 'Gentoo' and not salt.utils.systemd.booted(__context__):
         return __virtualname__
     return (False, 'The gentoo_service execution module cannot be loaded: '
-            'only available on Gentoo systems.')
+            'only available on Gentoo systems and/or systemd running.')
 
 
 def get_enabled():
```

**Comment by alexminder** [comment]
It looks strange for me, that you can't reproduce the issue.
I had check the history and found the issue has been fixed in the latest release by https://github.com/saltstack/salt/pull/36010 .

**Comment by gtmanfred** [comment]
Hrm, that is strange, I do not know why I was unable to reproduce it.  I only installed with emerge, and looking in the package source, I do not see that patch applied to salt.

This bug should be fixed in 2016.3.4 which is out now.

**Comment by alexminder** [comment]
Yep.
I see the issue fixed in 2016.3.4

**Comment by gtmanfred** [comment]
Awesome! thanks for confirming!
